test/reduce.js

```diff
@@ -0,0 +1,386 @@
+// tree and the last known good tree produce the same non-nil error-free output
+// after being run, then the permutation survives to the next generation and
+// is the basis for subsequent iterations.  The test case is reduced as a
+// consequence of complex expressions being replaced with simpler ones.
+// This function assumes that the testcase will not result in a parse or
+// runtime Error.  Note that a reduced test case will have different runtime
+// output - it is not functionally equivalent to the original. The only criteria
+// is that once the generated reduced test case is run without minification, it
+// will produce different output from the code minified with `minify_options`.
+// Returns a `minify` result object with an additonal boolean property `reduced`.
+
+module.exports = function reduce_test(testcase, minify_options, reduce_options) {
+    minify_options = minify_options || { compress: {}, mangle: false };
+    reduce_options = reduce_options || {};
+    var timeout = 1500; // start with a low timeout
+    var max_iterations = reduce_options.max_iterations || 1000;
+    var verbose = reduce_options.verbose;
+    var minify_options_json = JSON.stringify(minify_options);
+    var reduced = false;
+
+    if (testcase instanceof U.AST_Node) testcase = testcase.print_to_string();
+
+    // the initial timeout to assess the viability of the test case must be large
+    if (producesDifferentResultWhenMinified(testcase, minify_options, 5000)) {
+        // Replace expressions with constants that will be parsed into
+        // AST_Nodes as required.  Each AST_Node has its own permutation count,
+        // so these replacements can't be shared.
+        // Although simpler replacements are generally faster and better,
+        // feel free to experiment with a different replacement set.
+        var REPLACEMENTS = [
+            // "null", "''", "false", "'foo'", "undefined", "9",
+            "1", "0",
+        ];
+
+        // There's a relationship between each node's _permute counter and
+        // REPLACEMENTS.length which is why fractional _permutes were needed.
+        // One could scale all _permute operations by a factor of `steps`
+        // to only deal with integer operations, but this works well enough.
+        var steps = 4;        // must be a power of 2
+        var step = 1 / steps; // 0.25 is exactly representable in floating point
+
+        var tt = new U.TreeTransformer(function(node, descend, in_list) {
+            if (CHANGED) return;
+
+            // quick ignores
+            if (node instanceof U.AST_Toplevel) return;
+            if (node instanceof U.AST_Directive) return;
+
+            // technically these types should not have to be excluded
+            // from being dropped, but uglify will not correctly minify
+            // some non-strict generated reduced test cases.
+            if (node instanceof U.AST_Var) return;
+            if (node instanceof U.AST_VarDef) return;
```

Comment:
fyi... try commenting out these lines in your local repo and test ufuzz with an error introduced in lib/compress.js like the patch I tested with.  I don't know whether these test cases violate an uglify-js assumption, or whether it's a genuine bug.

Reply:
see https://github.com/mishoo/UglifyJS2/pull/3705#issuecomment-582243284

Reply:
I can't seem to reproduce this now that I rebased with latest master. Just to clarify, it used to generate test cases that Node could run correctly, but the latest release of uglify-js would minify incorrectly (not the local version with the introduced compressor error). Feel free to remove these lines in the future and see what happens. All this stuff is trial and error.

Reply:
I decided to comment out these lines. You can reintroduce them if there's a problem. reduce_test appears to produce smaller test cases as result and runs faster.